snuba/utils/streams/dummy.py

```diff
@@ -0,0 +1,106 @@
+from datetime import datetime
+from typing import (
+    Callable,
+    Mapping,
+    MutableMapping,
+    MutableSequence,
+    Optional,
+    Sequence,
+)
+
+from snuba.utils.streams.consumer import Consumer
+from snuba.utils.streams.types import Message, Partition, Topic, TPayload
+
+
+epoch = datetime(2019, 12, 19)
+
+
+class DummyConsumer(Consumer[TPayload]):
+    def __init__(self, messages: Mapping[Partition, Sequence[TPayload]]) -> None:
+        # TODO: The message data needs to include the timestamp.
+        self.__messages: Mapping[Partition, MutableSequence[TPayload]] = {
+            partition: list(payloads) for partition, payloads in messages.items()
+        }
+
+        self.__subscription: Sequence[Topic] = []
+        self.__assignment: Optional[Sequence[Partition]] = None
+
+        self.__offsets: MutableMapping[Partition, int] = {}
+        self.__staged_offsets: MutableMapping[Partition, int] = {}
+        self.__committed_offsets: MutableMapping[Partition, int] = {}
+
+        self.commit_offsets_calls = 0
+        self.close_calls = 0
+
+        self.__closed = False
+
+    def extend(self, messages: Mapping[Partition, Sequence[TPayload]]) -> None:
+        for partition, payloads in messages.items():
+            self.__messages[partition].extend(payloads)
```

Comment:
I don't think you are ever creating a new entry in self.__messages to extend after initializing the attribute in the constructor. Upon subscription you may be assigned a new partition.
Or is self.__messages meant to represent all partitions and messages no matter what has been assigned to this consumer, thus you would never add a new partition ?

Reply:
> I don't think you are ever creating a new entry in self.__messages to extend after initializing the attribute in the constructor. Upon subscription you may be assigned a new partition. Or is self.__messages meant to represent all partitions and messages no matter what has been assigned to this consumer, thus you would never add a new partition ?

Yeah — right now this just assumes that:

1. the number of partitions of partitions per topic is constant, and established in the constructor
2. this only emulates the behavior or a single consumer, not balanced consumption over a consumer group
3. those two assumptions lead to a third, which is that a consumer is always assigned the full set of partitions (because there are no other consumers) that are defined at startup (because there is no other way to add or remove partitions)

I'd like to eventually improve this so that things like rebalancing and partition addition can be tested (since these scenarios are kind of a pain to create in an integration test and our current coverage is pretty bad), but I think that will be a follow up that is broader in scope.